ripple/src/main/webapp/businesscarousels.js

```diff
@@ -75,52 +76,169 @@ function cardFunctionality() {
+    addDynamicCarouselByTag("black-owned-businesses", "Black-owned");
+    addDynamicCarouselByTime("under-20-mins-away", maxWalkingTime);
+    addDynamicCarouselByTag("trending-near-you", "Trending");
+    addDynamicCarouselByTag("up-and-coming", "New");
 }
 
-/* Dynamically loads content into Bootstrap carousel by reconstructing HTML elements and making a Firestore query. */
-function addDynamicCarousel(carouselId, tag) {
+/* Dynamically loads content into Bootstrap carousel by reconstructing HTML elements and making a Firestore query. 
+   Filters by tag.*/
+function addDynamicCarouselByTag(carouselId, tag) {
     db.collection("businesses").get().then((querySnapshot) => {
-        var contentStrings = [];
         var makeElement = false;
+        var firstCard = true;
         querySnapshot.forEach((doc) => {
-            //Check if the city in Firestore matches the city extracted from the user inputted address.
-            //Also checks if the business's tag list contains the tag that the carousel requires.
+            // Check if the city in Firestore matches the city extracted from the user inputted address.
+            // Also checks if the business's tag list contains the tag that the carousel requires.
             if (doc.data().address != null && doc.data().address[1] == localStorage.getItem('enteredCity') 
                 && doc.data().tags[convertToRawString(tag)] == tag) {
               makeElement = true;
               const card = document.createElement('div');
               card.classList = 'carousel-inner row w-100 mx-auto';
+              
+              // Distance matrix calculations
+              // Get user's latitude and longitude from localStorage
+              var userLat = localStorage.getItem('enteredLat');
+              var userLong = localStorage.getItem('enteredLong');
+
+              // Get business latitude and longitude from GeoPoint object in firestore
+              var busLat = doc.data().coordinates.latitude;
+              var busLong = doc.data().coordinates.longitude;
 
-            // Construct card content by appending HTML strings
-              var content = `
-                <div class="carousel-item col-md-4">
-                <div class="card">
-                    <img class="card-img-top img-fluid" id="card-dynamic-image" src="/serve?blob-key=${doc.data().thumbnailImage}">
-                    <div class="card-body">
-                    <p class="card-text">${doc.data().businessName[1]}</p>
-                    <p class="card-text"><small class="text-muted">${doc.data().walkingDistance}</small></p>
+              // Reconstruct as maps LatLng object
+              var origin = new google.maps.LatLng(userLat, userLong);
+              var destination = new google.maps.LatLng(busLat, busLong);
+              
+              // Call distance matrix service
+              var service = new google.maps.DistanceMatrixService();
+              service.getDistanceMatrix(
+                {
```

Comment:
nit: I feel like for passing in an object inline, you don't need to start the open bracket on the new line. So that the indentation for the content can be reduced.

Reply:
fixed